src/main/scala/Head.scala

```diff
@@ -1,5 +1,27 @@
 package reinvent
 
 object Head {
-  def apply(input: List[Int]): Int = ???
+  def apply[A](input: List[A]): A = input match {
+    case Nil => throw new RuntimeException("Head of empty list")
+    case (h::_) => h
+  }
 }
+
+object Filter {
+  def apply[A](items: List[A], predicate: A => Boolean): List[A] = {
+    def applyFilter(items: List[A], accumulator: List[A]) : List[A] = {
+
+      if (items.isEmpty)  return accumulator
+
+      else if (predicate(items.head)) {
+        applyFilter(items.tail, accumulator :+ items.head)
```

Comment:
If you're dealing with very large collections you may want to prepend and then reverse at the end.
This can actually work out faster (every time you append it must scan to the end of the collection).